### PR TITLE
Expectation examples

### PR DESCRIPTION
**Pull request by wiseodd**
Example of how to leverage the newly added `functional_samples` method #243. 

I also added the abstract interface for `samples` and `functional_samples` to `BaseLaplace`.

### CODE REVIEW

File: laplace/baselaplace.py [hunk 1]
```diff
@@ -694,6 +694,62 @@ def _glm_forward_call(
                 "Prediction path invalid. Check the likelihood, pred_type, link_approx combination!"
             )
 
+    def sample(
```
Comment (runame) [inline laplace/baselaplace.py:697]: Random thought and I probably don't remember enough context, but it seems like it would make sense for `BaseLaplace` to inherit from `abc.ABC` and use `@abc.abstractmethod` for cases like this one.

File: examples/expectation_example.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    print(f"Thompson sampling, {pred_type}")
+
+    x_test = torch.randn(10, 2)
+    x_test.requires_grad = True
+
+    f_sample = la.functional_samples(x_test, pred_type=pred_type, n_samples=1)
+    f_sample = f_sample.squeeze(0)  # We only use a single sample
+    print(f"TS shape: {f_sample.shape}, TS requires grad: {f_sample.requires_grad}")
+
+    grad_x = autograd.grad(f_sample.sum(), x_test)[0]
```
Comment (runame) [inline examples/expectation_example.py:47]: A comment describing what is happening here would be nice.

File: examples/expectation_example.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+    print(
+        f"Grad x_test shape: {grad_x.shape}, Grad x_test vanishing: {torch.allclose(grad_x, torch.tensor(0.0))}"
+    )
+    print()
+
+print()
+
+# Monte-Carlo expected improvement E_{f(x) ~ p(f(x) | D)} [max(f(x) - best_f, 0)]
```
Comment (runame) [inline examples/expectation_example.py:56]: Introduce later used abbreviation `(IE)` here.

File: examples/expectation_example.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    x_test = torch.randn(10, 2)
+    x_test.requires_grad = True
+
+    f_samples = la.functional_samples(x_test, pred_type=pred_type, n_samples=10)
+    ei = (f_samples - f_best.reshape(1, 1, 1)).clamp(0.0).mean(0)
+    print(f"EI shape: {ei.shape}, EI requires grad: {ei.requires_grad}")
+
+    grad_x = autograd.grad(ei.sum(), x_test)[0]
+
+    print(
```
Comment (runame) [inline examples/expectation_example.py:71]: Same here, adding a few comments describing what you are trying to show would be nice.

File: docs/expectation_example.md [hunk 5]
```diff
@@ -0,0 +1,120 @@
+## Motivation
```
Comment (runame) [inline docs/expectation_example.md:1]: I have looked at this file after `expectation_example.py`, maybe it is worth it to just copy some of the commentary here to the runnable example (as comments)?

I haven't thought about it much, but there should be a nicer solution than duplicating the example code in a markdown file for the docs and a runnable Python file. Maybe a Jupyter notebook that can be rendered as a markdown file?

**Review by runame — APPROVED** [review]
Besides my comments, LGTM.

File: docs/expectation_example.md (hunk 5 above)
Comment (aleximmer) [reply docs/expectation_example.md:1]: I agree the duplicated code is not great but I think it's already a bit longer like this. We should add an issue to write a small script that extracts python blocks from the md doc files and runs that. I think we introduced having them both as py and md files so we can make sure the examples actually run. Will add an issue for this.

File: laplace/baselaplace.py (hunk 1 above)
Comment (aleximmer) [reply laplace/baselaplace.py:697]: Agree, I think we use the old "implicit" way everywhere in the code at the moment though. So this should be done separately across the whole lib.

File: uv.lock [hunk 6]
```diff
@@ -601,7 +601,7 @@ wheels = [
 
 [[package]]
 name = "laplace-torch"
-version = "0.2.2"
+version = "0.2.2.2"
```
Comment (aleximmer) [inline uv.lock:604]: Why add another separator instead of going to `.3` here?

**Review by aleximmer — APPROVED** [review]
LGTM, thanks for adding the example. Some comment on a higher level but not necessary to add: it would be cool to plot the EI function and then compute the gradient at some point and add that to the plot as well to visualize the example mot intuitively.

File: laplace/baselaplace.py (hunk 1 above)
Comment (wiseodd) [reply laplace/baselaplace.py:697]: Let's do this in a new PR https://github.com/aleximmer/Laplace/issues/275

File: docs/expectation_example.md (hunk 5 above)
Comment (wiseodd) [reply docs/expectation_example.md:1]: https://github.com/aleximmer/Laplace/issues/276
